pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ChatCompletionsStep.java

```diff
@@ -0,0 +1,120 @@
+    JsonRecord jsonRecord = new JsonRecord();
+    if (transformContext.getKeySchema() != null) {
+      jsonRecord.setKey(
+          toJsonSerializable(transformContext.getKeySchema(), transformContext.getKeyObject()));
+    } else {
+      jsonRecord.setKey(transformContext.getKey());
+    }
+    jsonRecord.setValue(
+        toJsonSerializable(transformContext.getValueSchema(), transformContext.getValueObject()));
+    jsonRecord.setDestinationTopic(transformContext.getOutputTopic());
+
+    jsonRecord.setProperties(transformContext.getOutputProperties());
+    Record<?> currentRecord = transformContext.getContext().getCurrentRecord();
+    currentRecord.getEventTime().ifPresent(jsonRecord::setEventTime);
+    currentRecord.getTopicName().ifPresent(jsonRecord::setTopicName);
+
+    List<ChatMessage> messages =
+        config
+            .getMessages()
+            .stream()
+            .map(
+                message ->
+                    new ChatMessage(message.getRole())
+                        .setContent(applyTemplate(message.getContent(), jsonRecord)))
+            .collect(Collectors.toList());
+
+    ChatCompletionsOptions chatCompletionsOptions =
+        new ChatCompletionsOptions(messages)
+            .setMaxTokens(config.getMaxTokens())
+            .setTemperature(config.getTemperature())
+            .setTopP(config.getTopP())
+            .setLogitBias(config.getLogitBias())
+            .setUser(config.getUser())
+            .setStop(config.getStop())
+            .setPresencePenalty(config.getPresencePenalty())
+            .setFrequencyPenalty(config.getFrequencyPenalty());
+
+    ChatCompletions chatCompletions =
+        client.getChatCompletions(config.getDeploymentId(), chatCompletionsOptions);
+
+    String content = chatCompletions.getChoices().get(0).getMessage().getContent();
+    // TODO: At the moment, we only support outputing the value with STRING schema.
+    transformContext.setValueSchema(Schema.STRING);
+    transformContext.setValueObject(content);
+  }
+
+  private static Object toJsonSerializable(Schema<?> schema, Object val) {
+    if (schema == null || schema.getSchemaInfo().getType().isPrimitive()) {
+      return val;
+    }
+    switch (schema.getSchemaInfo().getType()) {
+      case AVRO:
+        ObjectMapper mapper = new ObjectMapper();
+        // TODO: do better than the double conversion AVRO -> JSON -> Map
+        return mapper.convertValue(
+            JsonConverter.toJson((org.apache.avro.generic.GenericRecord) val),
+            new TypeReference<Map<String, Object>>() {});
+      case JSON:
+        return val;
+      default:
+        throw new UnsupportedOperationException(
+            "Unsupported schemaType " + schema.getSchemaInfo().getType());
+    }
+  }
+
+  private String applyTemplate(String template, JsonRecord record) {
+    // TODO: pre-compile the templates at init time
+    Template tmpl = Mustache.compiler().compile(template);
```

Comment:
Is it possible to extend the Mustache template to leverage `JstlTransformContextAdapter`? This would enable users to use the `fn:` functions or evaluate any field the way compute does it. If not, at least we need to make sure that the `dot` notation to access fields is consistent with compute.

Reply:
this is a great idea

Reply:
Mustache uses the `dot` annotation too.
I don't see how it can be mixed with JstlTransformContextAdapter

Reply:
I was thinking if Mustache allows for custom evaluations so first level is to detect anything between `{{}}` and the second step is to delegate to `JstlTransformContextAdapter` - on the second thought this may confuse the user because the syntax will be mixed between the two - maybe better to stick with Mustache